engineapi/proxy.go

```diff
@@ -0,0 +1,299 @@
+package engineapi
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	imclient "github.com/longhorn/longhorn-instance-manager/pkg/client"
+	imtypes "github.com/longhorn/longhorn-instance-manager/pkg/types"
+	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+)
+
+const (
+	ProxyPollInterval = 5 * time.Second
+)
+
+type EngineClientProxy interface {
+	EngineClient
+	BackupTargetBinaryClient
+
+	IsGRPC() bool
+	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
+	Stop(string) error
+	Ping() error
+}
+
+type EngineClientProxyHandler struct {
+	Clients map[string]EngineClientProxy
+
+	logger logrus.FieldLogger
+
+	ds *datastore.DataStore
+
+	lock *sync.RWMutex
+}
+
+func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore) *EngineClientProxyHandler {
+	return &EngineClientProxyHandler{
+		Clients: make(map[string]EngineClientProxy),
+		logger:  logger,
+		ds:      ds,
+		lock:    &sync.RWMutex{},
+	}
+}
+
+func getLoggerForProxy(logger logrus.FieldLogger, image string) *logrus.Entry {
+	return logger.WithFields(
+		logrus.Fields{
+			"instanceManagerImage": image,
+		},
+	)
+}
+
+func (h *EngineClientProxyHandler) GetClient(im *longhorn.InstanceManager, fallBack interface{}) (c EngineClientProxy, err error) {
+	if im == nil {
+		return nil, errors.Errorf("BUG: failed to get proxy client due to missing instance manager")
+	}
+
+	log := getLoggerForProxy(h.logger, im.Spec.Image)
+
+	shouldFallBack := false
+
+	if im != nil {
+		if err := CheckInstanceManagerProxyCompatibility(im); err != nil {
+			log.WithError(err).Warnf("Use fallback client")
+			shouldFallBack = true
+		}
+	}
+
+	if shouldFallBack {
+		if fallBack == nil {
+			return nil, errors.Errorf("missing fallback client for proxy")
+		}
+
+		if obj, ok := fallBack.(*EngineBinary); ok {
+			return obj, nil
+		}
+
+		if obj, ok := fallBack.(*BackupTargetClient); ok {
+			return obj, nil
+		}
+
+		return nil, errors.Errorf("BUG: invalid fallback client for proxy: %v", fallBack)
+	}
+
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	client, exist := h.Clients[im.Spec.Image]
+	if exist {
+		err := client.Ping()
```

Comment:
Should this ping function be protected by the h.lock?

Reply:
I've put the lock around delete and create of the client. WDYT?

Reply:
Need to protect the read as well:
```
h.lock.Lock()
client, exist := h.Clients[im.Spec.Image]
h.lock.Unlock()
```

Reply:
h.lock is a RWMutex. Why not use RLock/RUnLock for read operation?
```
h.lock.RLock()
client, exist := h.Clients[im.Spec.Image]
h.lock.RUnlock()
```